### PR TITLE
feat: implement dropping building blocks on canvas

### LINKED ISSUE #31856: [Feature]: Implement logic for dropping building blocks in the canvas
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Summary
The objective of this feature request is to enhance user experience by enabling the dragging and dropping of building blocks onto the canvas, similar to widgets. This functionality will streamline the process of integrating building blocks into the user's application.

Proposal
We propose implementing a skeleton loader widget that appears after the user releases a building block onto the canvas, when all API calls and processes are complete we will use the existing copy and paste logic to drop the widgets into the canvas for the user.

### Why should this be worked on?

it is needed to complete the implementation of building blocks on canvas project.

### PR DESCRIPTION
**Pull request by jacquesikot**
**Goal**
To drag any building blocks from the explorer and drop at any position on the users canvas with full implementation.

**Implementation**
- Show skeleton widget when building block is initially dragged unto the canvas.
- Make API call to add datasources, queries, and JS to existing page of users app
- Get returned widget DSL and use existing copy paste logic to display widgets on the canvas
- Remove loading state, handle clean up for copy paste
- Run all queries newly created by the dropped building block

**Limitations**
- There is a loading state and the process is not instant like dropping a widget

This PR is followed by this one [here](https://github.com/appsmithorg/appsmith/pull/31406), which displays the loading state when a building block is dragged unto the canvas.

Fixes #31856
## Automation

/ok-to-test tags="@tag.Templates, @tag.MainContainer, @tag.Visual, @tag.Widget"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/8701664455>
> Commit: 456a7a0a322e76974a7f5c41a6c1e274ef82e4ea
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=8701664455&attempt=1" target="_blank">Click here!</a>

## Summary by CodeRabbit

- **New Features**
    - Introduced a new "Building Blocks" category in the widget sidebar for enhanced organization and accessibility.
    - Added functionality to resize building blocks with new horizontal and vertical limits.
    - Implemented a "see more" button for "Building Blocks" to display all associated widgets.
    - Enhanced drag and drop functionality for building blocks on the canvas.

- **Enhancements**
    - Improved sorting logic for widgets, prioritizing "Building Blocks".
    - Enhanced widget search functionality within the sidebar.

- **Bug Fixes**
    - Adjusted default rows and columns settings for explorer building blocks to improve layout and usability.

- **Documentation**
    - Updated messages and constants related to new features for clarity and consistency.

- **Refactor**
    - Refactored drag and drop handling logic to support new building block constraints and features.
    - Updated application state management to include building blocks related data.

### CODE REVIEW

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 1]
```diff
@@ -496,7 +545,114 @@ function* addUIEntitySaga(addEntityAction: ReduxAction<WidgetAddChild>) {
...
+
+      const response: ApiResponse = yield call(
+        ApplicationApi.importBuildingBlockToApplication,
+        body,
+      );
+      const isValid: boolean = yield validateResponse(response);
+
+      if (isValid) {
+        const buildingBlockDsl = JSON.parse(response.data as string);
+        const buildingBlockWidgets = buildingBlockDsl.children;
```
Comment (rahulbarwal) [inline app/client/src/sagas/WidgetAdditionSagas.ts:570]: Refactor the putting BB widgets on UI to a separate function.

File: app/client/src/ce/api/ApplicationApi.tsx [hunk 2]
```diff
@@ -267,6 +268,18 @@ export interface ImportBuildingBlockToApplicationRequest {
...
+export interface ImportBuildingBlockToApplicationResponse {
+  widgetDsl: string;
+  onPageLoadActions: {
+    confirmBeforeExecute: boolean;
+    id: string;
+    jsonPathKeys: string[];
+    name: string;
+    pluginType: PluginType;
+    timeoutInMillisecond: number;
+  }[];
```
Comment (rahulbarwal) [inline app/client/src/ce/api/ApplicationApi.tsx:280]: Do we not have an interface for it already?

File: app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx [hunk 3]
```diff
@@ -72,6 +74,9 @@ const StyledDropTarget = styled.div`
 
 function Onboarding() {
   const isMobileCanvas = useSelector(getIsMobileCanvasLayout);
+  const draggingBuildingBlockToCanvas = useSelector(
```
Comment (rahulbarwal) [inline app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx:77]: variable name does not tell us if it is a boolean. Consider adding a `is` to beginning.

File: app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx [hunk 4]
```diff
@@ -97,15 +102,31 @@ function Onboarding() {
     ],
   );
 
-  if (shouldShowStarterTemplates && appState === IDEAppState.EDITOR)
+  const shouldShowBuildingBlocksDropTarget = useMemo(
+    () => releaseDragDropBuildingBlocks && !draggingBuildingBlockToCanvas,
+    [releaseDragDropBuildingBlocks, draggingBuildingBlockToCanvas],
+  );
+
+  if (shouldShowStarterTemplates && appState === IDEAppState.EDITOR) {
```
Comment (rahulbarwal) [inline app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx:110]: This check does not consider if drag drop on BB is enabled, I guess that's why we are seeing BB on canvas with this project.

File: app/client/src/layoutSystems/common/utils/canvasDraggingUtils.ts [hunk 5]
```diff
@@ -465,10 +466,11 @@ export const modifyBlockDimension = (
   parentBottomRow: number,
   canExtend: boolean,
   modifyBlock: boolean,
+  horizontalMinResizeLimit: number = HORIZONTAL_RESIZE_MIN_LIMIT,
+  verticalMinResizeLimit: number = VERTICAL_RESIZE_MIN_LIMIT,
```
Comment (rahulbarwal) [inline app/client/src/layoutSystems/common/utils/canvasDraggingUtils.ts:470]: Add some comment here to preserve the context for future development.

**Review by rahulbarwal — COMMENTED** [review]
Requesting minor changes. get your backend changes in this PR reviewed as well.
@marks0351 Can you please review this as well please?

PS: Smaller PRs please.

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 6]
```diff
@@ -480,23 +518,199 @@ function* addNewTabChildSaga(
...
+            };
+          },
+        ),
+      );
+      yield saveCopiedWidgets(
+        JSON.stringify({
+          widgets: widgetsToPasteInCanvas,
+          flexLayers: [],
+        }),
+      );
```
Comment (rahulbarwal) [inline app/client/src/sagas/WidgetAdditionSagas.ts:604]: What if user already has something copied?
We have to preserve that as well. What it means is, cache the existing value and put it back again, once our work is done.

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 7]
```diff
@@ -480,23 +518,199 @@ function* addNewTabChildSaga(
...
+
+      const widgetsToPasteInCanvas: CopiedWidgetData[] = yield all(
+        flattenedBlockWidgets.map(
+          (widget: FlattenedWidgetProps, index: number) => {
+            let widgetPositionInfo: WidgetLayoutPositionInfo | null = null;
+            if (
+              widget.parentId &&
+              layoutSystemType === LayoutSystemTypes.ANVIL
+            ) {
```
Comment (rahulbarwal) [inline app/client/src/sagas/WidgetAdditionSagas.ts:578]: Why do we have this condition? are we supporting anvil?

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 8]
```diff
@@ -480,23 +518,199 @@ function* addNewTabChildSaga(
...
-        ...addEntityAction,
+    const dragDetails: DragDetails = yield select(getDragDetails);
+    const buildingblockName = dragDetails.newWidget.displayName;
+    const buildingBlocks: Template[] = yield select(getTemplatesSelector);
+    const applicationId: string = yield select(getCurrentApplicationId);
+    const currentPageId: string = yield select(getCurrentPageId);
+    const workspaceId: string = yield select(getCurrentWorkspaceId);
+    const allWidgets: CanvasWidgetsReduxState = yield select(getWidgets);
+    const existingCopiedWidgets: unknown = yield call(getCopiedWidgets);
+    const canvasWidgets: CanvasWidgetsReduxState =
```
Comment (rahulbarwal) [inline app/client/src/sagas/WidgetAdditionSagas.ts:535]: What is the diff between all widgets and canvasWidgets?Should these not be same?

File: app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx (hunk 4 above)
Comment (marks0351) [inline app/client/src/layoutSystems/common/dropTarget/DropTargetComponent.tsx:110]: could you add more meaningful variables to each of the if case conditions to make it more readable.

File: app/client/src/layoutSystems/common/dropTarget/buildingBlockExplorerDropTarget/index.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
+import { importSvg } from "@design-system/widgets-old/src/utils/icon-loadables";
+import { Text } from "design-system";
+import React from "react";
+import styled from "styled-components";
+
+export const DropTargetContainer = styled.div`
```
Comment (marks0351) [inline app/client/src/layoutSystems/common/dropTarget/buildingBlockExplorerDropTarget/index.tsx:6]: can you add semantic constants for the values like top, left, padding, margin-left.
I want to understand how we arrived at these values.

File: app/client/src/pages/Editor/widgetSidebar/hooks.ts [hunk 10]
```diff
@@ -16,9 +16,8 @@ import { groupWidgetCardsByTags } from "../utils";
  * @returns Object containing grouped cards and cards.
  */
 export const useUIExplorerItems = () => {
-  const releaseDragDropBuildingBlocks = useFeatureFlag(
-    FEATURE_FLAG.release_drag_drop_building_blocks_enabled,
-  );
+  const releaseDragDropBuildingBlocks = true;
```
Comment (marks0351) [inline app/client/src/pages/Editor/widgetSidebar/hooks.ts:19]: Are these changes for dev purpose?

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 11]
```diff
@@ -480,23 +518,199 @@ function* addNewTabChildSaga(
...
+    });
+
+    // makes sure updateAndSaveLayout completes first for skeletonWidget addition
+    yield take(ReduxActionTypes.SAVE_PAGE_SUCCESS);
+
+    // api call adds DS, queries and JS to page and returns new page dsl with building block
+    const response: ApiResponse<ImportBuildingBlockToApplicationResponse> =
+      yield call(ApplicationApi.importBuildingBlockToApplication, body);
+
+    const isValid: boolean = yield validateResponse(response);
```
Comment (marks0351) [inline app/client/src/sagas/WidgetAdditionSagas.ts:562]: Would recommend breaking this function, it is doing so many things.
- API resource fetch and validation
- perform a psuedo Copy 
- select the position to paste
- delete the skeleton widget
- perform a psuedo Paste
- save the layout

Lets create a single purpose functions, tweak the implementation and make it atomic.

- sepearate function to fetch BB resource and validate it and return the results
- compute the widgets state where skeleton widget is deleted using getUpdatedDslAfterDeletingWidget
- select position to place the widget
- use a extracted version of pastesagas which will just paste widgets in a positon when given a canvasWidgets State.
- save the layout.

WDYT?

**Review by marks0351 — CHANGES_REQUESTED** [review]
I also see that building blocks when added to a container are added back to main_container. pls do check it.

**Comment by marks0351** [comment]
@jacquesikot I would also recommend to restrict Building blocks to show up only for fixed layout.

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 12]
```diff
@@ -480,23 +527,239 @@ function* addNewTabChildSaga(
...
+    skeletonWidget.widgetId,
+  );
+}
+
+function* addUIEntitySaga(addEntityAction: ReduxAction<WidgetAddChild>) {
+  try {
+    const { payload } = addEntityAction;
+    const { type } = payload;
+
+    if (type === BUILDING_BLOCK_EXPLORER_TYPE) {
```
Comment (marks0351) [inline app/client/src/sagas/WidgetAdditionSagas.ts:761]: why don't we dispatch a different redux action of building block addtion action?

**Review by marks0351 — CHANGES_REQUESTED** [review]

**Pull request merged**